esp-hal/src/dma/mod.rs

```diff
@@ -732,58 +730,65 @@ where
     }
 
     fn available(&mut self) -> usize {
-        if self.last_seen_handled_descriptor_ptr.is_null() {
-            self.last_seen_handled_descriptor_ptr = self.descriptors.as_mut_ptr();
-            return 0;
-        }
-
-        if self.available != 0 {
-            return self.available;
-        }
-
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let mut last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
 
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
-
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
-
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
 
+            last_in_descr.set_owner(Owner::Dma);
+            last_in_descr.set_suc_eof(false);
             unsafe {
-                descr_address.write_volatile(dw0);
+                last_in_dscr_ptr.write_volatile(last_in_descr);
             }
 
-            self.last_seen_handled_descriptor_ptr = if next_descriptor.is_null() {
-                self.descriptors.as_mut_ptr()
-            } else {
-                next_descriptor
-            };
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+                return self.available;
+            }
```

Comment:
This is technically an existing bug but a bug none the less.

Isn't it incorrect here to give ownership of the descriptor (and buffer it points to) to the hardware before the application is done reading from it?
I think the `pop` method should do the work of `.set_owner(Owner::Dma);` since that's when the buffer is no longer read by the cpu/application.

Reply:
Good idea - I agree